### ISSUE TITLE
PMD custom rule set documentation unclear

### ISSUE THREAD

**Issue by jordanjennings** [gradle/gradle#8514]
### Expected Behavior
When I override the PMD configuration with my own rule set file, it should only use the rules in that file.

### Current Behavior
The rules in my file seem to be added to the default rules that the plugin applies.

After much frustration (and lots of time) I eventually found this line in `PmdExtension` javadocs:

>If you want to only use custom rule sets, you must clear {@code ruleSets}.

(found here: https://github.com/gradle/gradle/blob/master/subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdExtension.java#L167)

This note does not make it into any HTML documentation. On the plugin page it should clearly call out how to override the configuration with an XML file and that that if I provide my own configuration file I need to manually set `ruleSets = []`.

I'd argue that it should internally set `ruleSets = []` if a custom rule set is provided, but I understand that would be a breaking change. For now the docs should be updated, and perhaps the design revisited when appropriate.

### Context
Trying to apply the PMD plugin with a customized `ruleset.xml`.

### Steps to Reproduce (for bugs)
Use a custom ruleset with the following configuration:

```
pmd {
  ruleSetConfig = resources.text.fromFile('ruleset.xml')
  consoleOutput = true
}
```

### Your Environment
Gradle 5.2.1
OpenJDK 11.0.1
Ubuntu 18.04

**Comment by Vampire** [comment]
I second that request and was about to report it myself.
Especially as you might not recognize this behavior instantly.
If you still use Gradle 4 but with a recent PMD version, you see it, as you get the warning about the deprecated `rulesets/java/basic.xml` names in that case.
But if you are on Gradle 5 or use the default PMD version in Gradle 4 you wouldn't instantly recognize that there are additional rules applied unless they trigger and you wonder why those are included in the report when you didn't configure them.

I suggest to deprecate this behavior by logging a deprecation warning when custom ruleset file(s) is/are set and the ruleSets property is not set explicitly (to some value, to empty list or `.clear()` called) and then in the next major version have the convention mapping be empty if custom ruleset files are set. If one wants to configure it both, via `ruleSets` and in a custom file, it should still be possible explicitly probably.

**Comment by Vampire** [comment]
Still the same in 7.4.2, please reopen

**Comment by cvmocanu** [comment]
Doing the recommended `ruleSets.clear()` with Gradle 8 results in an exception:
```
java.lang.UnsupportedOperationException
        at com.google.common.collect.ImmutableCollection.clear(ImmutableCollection.java:346)
```

This is a breaking change, and a particularly bad one, since doing the recommended action results in breakage.

While I do think the change is for the better, this should be on the Gradle 8 release page, and in a prominent place.

**Comment by Vampire** [comment]
I'd suggest you open a fresh regression issue about it.
Then it will get more and immediate attention, compared to commenting on a closed issue. :-)

**Comment by cvmocanu** [comment]
Since it's a braking change, this ticket is incomplete without updating the release notes.
The ticket should be reopened.

**Comment by Vampire** [comment]
It is a breaking change and one that did not have a deprecation warning first.
Internally `ruleSets` previously was a `List` and now is a `ListProperty`.
Previously `getRuleSets` just returned that list, now it returns `ruleSets.get()` which is unmodifiable.
I still recommend you open a fresh regression ticket about it for the mentioned reasons.
There was another breaking change that was done without deprecation and that actually shouldn't have done.
I opened a regression ticket about it and it is already fixed in 8.0.1. ;-)
